x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateParse.java

```diff
@@ -122,7 +122,15 @@ public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEv
     }
 
     private static DateFormatter toFormatter(Object format, ZoneId zone) {
-        return forPattern(((BytesRef) format).utf8ToString()).withZone(zone);
+        String pattern = ((BytesRef) format).utf8ToString();
+        DateFormatter formatter;
+        try {
+            formatter = forPattern(pattern);
+        } catch (ClassCastException e) {
```

Comment:
This catch is too broad, try to make it more specific to the JDK bug.

Also, since the `forPattern` method comes from an ES class (which is used in many other places), why has this issue been visible in ESQL only? Or is not only ESQL, but other places? If there are other places where this behavior can be possible, how did we handle it?

Reply:
It seems to be extremely unlikely to happen: that API is used by tens of components in ES since 2018, none of them is explicitly managing this exception and I cannot find any bug reports about that.

I guess we caught it because we are running a significant amount of randomized tests that hit that API directly.

The root problem is [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/time/DateFormatters.java#L2126): `new DateTimeFormatterBuilder().appendPattern(input)`

Given the low practical impact I tried to keep the fix limited to ESQL, but I can move it to Server

Reply:
> It seems to be extremely unlikely to happen

On the contrary: because that API is heavily used in many components in ES, I find it very easy to encounter this JDK bug somewhere else. But it didn't happen so far. So what we do in ESQL or in how we test it, is very specific and different from the rest of Elasticsearch.

Look into why is that. See what is so specific about the test that fails. Knowing exactly why it fails for us and not for the rest of ES can reveal something that can be fixed/adjusted in some other place.

Reply:
I see your point, it makes sense, I'm moving the fix to server

> So what we do in ESQL or in how we test it, is very specific and different from the rest of Elasticsearch

> Look into why is that. See what is so specific about the test that fails. Knowing exactly why it fails for us and not for the rest of ES can reveal something that can be fixed/adjusted in some other place.

it fails for us because we [randomize](https://github.com/elastic/elasticsearch/pull/105048/files#diff-f5e8b736a246b4a56a08ef1e5e0b3646844db19feaafffe3c63b1d057cf79d9cL108) the input and we run this test four times per build (and still, it failed only once in nine months, since we introduced this function), while the rest of ES most of the times uses constant strings like `forPattern("yyyy-MM-dd")`